### PR TITLE
List Bot Tags On Site

### LINKED ISSUE #687: Page listing all tags and their content
I think it would be beneficial to have a page with toggle dropdowns listing all the tags we have and their contents for easy viewing.

We have the `!tags` on the server, but to read their contents you would have to run `!tag-name` on each individual tag or browse to our bot repository and read the markdown files directly. Both are not quite the smooth user experience I would want the folks on our server to have.

We can can build this page somewhat dynamically as it can pull from the specific bot repository on GitHub itself and use a jinja template to build the page from the markdown files.

### PR DESCRIPTION
**Pull request by HassanAbouelela**
Closes #687.

I implemented this feature in the most efficient way (in terms of network requests) possible. We download the entire bot project as a tarfile, extract all the tags from it, and cache them to the database.

Ideally, I'd also like to add a bit more flavor-info to the pages, such as when they were last updated, and the authors that have worked on them, but the only way to get that data from the API is to request data for a specific commit, which means making ~100 requests in total. We can't do that without being rate-limited, but it could be something to consider. I also considered downloading the git repo (as opposed to the content only), but it doesn't seem the API returns that, we'd need to clone with git.

Speaking of the 100 tags, I've gone through quite a few and they all look fine. They all should *render* fine, since the static deployment hits them all, but there could be some formatting oddities that I have missed. Feel free to tell me if so.

### CODE REVIEW

**Comment by jahrules** [comment]
> Closes #687.
> 
> I implemented this feature in the most efficient way (in terms of network requests) possible. We download the entire bot project as a tarfile, extract all the tags from it, and cache them to the database.
> 
> Ideally, I'd also like to add a bit more flavor-info to the pages, such as when they were last updated, and the authors that have worked on them, but the only way to get that data from the API is to request data for a specific commit, which means making ~100 requests in total. We can't do that without being rate-limited, but it could be something to consider. I also considered downloading the git repo (as opposed to the content only), but it doesn't seem the API returns that, we'd need to clone with git.
> 
> Speaking of the 100 tags, I've gone through quite a few and they all look fine. They all should _render_ fine, since the static deployment hits them all, but there could be some formatting oddities that I have missed. Feel free to tell me if so.

What about the tag "rule <#>"

**Comment by ChrisLovering** [comment]
> What about the tag "rule <#>"

Those aren't tags, they're a command in and of themselves. Also, the rules are already listed in our website (it's what the rules command links to).

File: pydis_site/apps/content/utils.py
Comment (ChrisLovering) [inline pydis_site/apps/content/utils.py]:
```suggestion
                content = content[:100] + "…"
```
😡

File: pydis_site/apps/content/views/tags.py
Comment (ChrisLovering) [inline pydis_site/apps/content/views/tags.py]: Would a `re.finditer` be more readable here? Rather than managing the iteration ourselves with `match.end()`.

File: pydis_site/apps/content/utils.py
Comment (ChrisLovering) [inline pydis_site/apps/content/utils.py]: I think we can get away with more than 100 chars in the preview, even on mobile, something ~300 is likely fine.
The problems arise when we start grabbing the first parts of a codeblock, and then we don't close them.
Is there a way for us to detect if we're "inside" a codeblock after trimming, and close it before the ellipses? 

Also, if the start of tags include a long hyperlink, the link length is counted towards this 100 character limit. This leads to some tags looking far shorter than other (example below). I wonder if there is anything we can do about this?

image

**Review by ChrisLovering — CHANGES_REQUESTED** [review]
Just a couple a comments here.

Another observation is that this list is _very_ long. A paginated and/or searchable list could go a long way here.
I don't think this should be a blocker on this PR directly, but it should be something we look into in future.

File: pydis_site/apps/content/utils.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/utils.py]: Hahaha, I'm actually working on making the cropping here better, just keep your eyes peeled.

File: pydis_site/apps/content/views/tags.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/views/tags.py]: Yes, that's what I originally did, but how do you do the substitution? Substituting by index would not work after the first match since the index would've moved, and substituting with something like re.sub wouldn't work either since the substitution still matches, so you'll end up on the first URL again.

File: pydis_site/apps/content/utils.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/utils.py]: It's not easy to fix this, at least in python, without effectively writing a full markdown interpreter. What I'm doing instead is re-writing this to do the splitting in javascript, where the text is already separated out in a convenient manner.

File: pydis_site/apps/content/views/tags.py
Comment (ChrisLovering) [reply pydis_site/apps/content/views/tags.py]: I made this suggestion without even looking at what this was doing. re.sub could be used here afaik, if you make the replacement a func rather than a string.
```py
def linkify_tag_ref(match: re.Match) -> str:
    link = reverse("content:tag", kwargs={"name": match.group("name")})
    return f"[{match.group()}]({link})"

...

content = COMMAND_REGEX.sub(linkify_tag_ref, content):
```

File: pydis_site/apps/content/models/tag.py
Comment (Numerlor) [inline pydis_site/apps/content/models/tag.py]: does this need to be its own field instead of using the name and a configured github url?

**Review by Numerlor — CHANGES_REQUESTED** [review]
It looks like this is currently missing proper support for tag groups (tags in directories under `tags/`), as there can be multiple tags with the same name under different dirs, and the github url template skips groups completely.

File: pydis_site/apps/content/models/tag.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/models/tag.py]: It's a remnant from when I was still using the API and it'd return the URL. If we leave things as is, I'll drop it, but it might come in handy for the groups.

**Comment by HassanAbouelela** [comment]
@Numerlor I was not aware of the tag groups, is there an example of them currently?

**Comment by Numerlor** [comment]
> @Numerlor I was not aware of the tag groups, is there an example of them currently?

I don't think any are currently used or that there are any real examples, but they're created through the directories I mentioned:
```
tags/
├── tag_group1/
│   └── tag.md
├── tag_group2/
│   └── another_tag.md
└── plain_tag_without_group.md
```
this structure would create two tag groups `tag_group1` and `tag_group2` with `tag.md` and `another_tag.md` in them respectively

**Comment by HassanAbouelela** [comment]
Thanks for the info, I'll work on implementing it

File: pydis_site/apps/content/views/tags.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/views/tags.py]: 74226511

File: pydis_site/apps/content/utils.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/utils.py]: f2ad3eed

File: pydis_site/apps/content/utils.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/utils.py]: Stale due to f2ad3eed

**Comment by HassanAbouelela** [comment]
Tag groups added in 45cdb27a. Due to the way tags are loaded in static previews, it was pretty easy to add in some under a fake group. Look for "very-cool-group" (these are sorted alphabetically like all the other tags, but we can place tag groups at the top of the list if we want to).

Live example: https://deploy-preview-763--pydis-static.netlify.app/pages/tags/very-cool-group/

File: pydis_site/apps/content/models/tag.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/models/tag.py]: Ended up making it a property in b4911d0

File: pydis_site/apps/content/migrations/0002_remove_tag_url_tag_group.py
Comment (ChrisLovering) [inline pydis_site/apps/content/migrations/0002_remove_tag_url_tag_group.py]: Do we really need a second migration file for un-merged changes?

Probably better to merge the two migrations together, so that there's only 1 migration to run after merge.

**Comment by Numerlor** [comment]
> Ideally, I'd also like to add a bit more flavor-info to the pages, such as when they were last updated, and the authors that have worked on them, but the only way to get that data from the API is to request data for a specific commit, which means making ~100 requests in total. We can't do that without being rate-limited, but it could be something to consider. I also considered downloading the git repo (as opposed to the content only), but it doesn't seem the API returns that, we'd need to clone with git.

This should be relatively simple to cache; the `contents` api on a directory returns all the files with their git blob SHA. If that is saved the individual requests for commits to a file can only be done if the new SHA differs from the saved one. Full content to the contents endpoint could also be avoided as it provides the ETag header but the download size there should be fairly small

**Comment by HassanAbouelela** [comment]
The cache solution sounds smart, I'll look into that. In terms of downloading, I'm not sure there's a good way to bulk download files without downloading the entire repo, did you have a solution in mind?

File: pydis_site/apps/content/migrations/0002_remove_tag_url_tag_group.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/migrations/0002_remove_tag_url_tag_group.py]: I figured it was easiest to write, and easiest to use for reviewers, but yeah we can merge them.

**Comment by Numerlor** [comment]
> The cache solution sounds smart, I'll look into that. In terms of downloading, I'm not sure there's a good way to bulk download files without downloading the entire repo, did you have a solution in mind?

I think the best solution is to just use the raw download urls from the contents api call. I don't believe count into the rate limits either so the only disadvantage is that is needs to be individual requests instead of bulk for the changed files, but that should only be a disadvantage when filling up the cache, as after that the changes will be mostly to individual files.

The GraphQL api can fetch all the files and their contents together, but I'm not sure if that can then be further filtered when not all files are necessary, and it needs auth

**Comment by HassanAbouelela** [comment]
I added in the commit stuff, I think it turned out decently.

File: pydis_site/apps/content/migrations/0002_remove_tag_url_tag_group.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/migrations/0002_remove_tag_url_tag_group.py]: f2374900c4c83097c105b56de02ea82d66bd9466

File: pydis_site/apps/content/models/tag.py
Comment (ChrisLovering) [inline pydis_site/apps/content/models/tag.py]: Possibly worth changing this to specify that it's a commit from the bot repo, since we're hard coding the URL

File: pydis_site/apps/content/models/tag.py
Comment (ChrisLovering) [inline pydis_site/apps/content/models/tag.py]: Is it worth pluralising this to make it clear it's a list of authors, rather than a single author. From the format_users function, it seems that this is a serialised list of dicts? If so, probably worth mentioning that in the help text too, maybe even a comment describing the structure, or a link to the docs.

File: pydis_site/apps/content/models/tag.py
Comment (ChrisLovering) [inline pydis_site/apps/content/models/tag.py]: Should this be format_authors, rather than format_users, so that we're not referring to it in 2 different ways

File: pydis_site/apps/content/models/tag.py
Comment (ChrisLovering) [inline pydis_site/apps/content/models/tag.py]: Since we're hard coding this to the bot repo, should this instead be
```suggestion
    """A tag from the python-discord bot repository."""
```

File: pydis_site/apps/content/models/tag.py
Comment (ChrisLovering) [inline pydis_site/apps/content/models/tag.py:58]: How can this be nullable?

File: pydis_site/apps/content/urls.py
Comment (ChrisLovering) [inline pydis_site/apps/content/urls.py]: Is this diff needed?

File: pydis_site/apps/content/utils.py
Comment (ChrisLovering) [inline pydis_site/apps/content/utils.py:20]: This is fine as is. However, what are your thoughts on using the TTLCache from [cachetools](https://pypi.org/project/cachetools/)? Just mentioning it was it's simply a deco around the function, as shown in the 3rd function [here](https://github.com/tkem/cachetools/#cachetools)

File: pydis_site/apps/content/views/tags.py
Comment (ChrisLovering) [inline pydis_site/apps/content/views/tags.py]: Tag groups are separated by spaces, rather than dashes. The below may work, but haven't tested if it causes any other oddities due to the space.
```suggestion
COMMAND_REGEX = re.compile(r"`*!tags? (?P<name>[\w\d ]+)`*")
```

**Review by ChrisLovering — CHANGES_REQUESTED** [review]
A couple of comments, only the regex one is required, the rest are just observations.

Haven't reviewed js, css or html directly, but it looks good in the app itself.

File: pydis_site/apps/content/models/tag.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/models/tag.py]: I called it author to follow the naming scheme of the API object, but it probably makes more sense to properly title it. In terms of the documentation, there's also the option of making the author another model, though it really only carries two important bits of info. Another option is to encode the author as a single string from the get-go if we wanted to avoid converting to and from JSON.

File: pydis_site/apps/content/models/tag.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/models/tag.py:58]: The initial value can be null, since the data is only populated when requesting a single tag page (to avoid making 100 requests at once). Once the tag is accessed for the first time (since the last update), it's commit data is saved for future access.

Comment (HassanAbouelela) [reply pydis_site/apps/content/models/tag.py:58]: Relevant code in the get_tag method:
https://github.com/python-discord/site/blob/92a42694b6ad1a29e5a21e0b3e57639528837113/pydis_site/apps/content/utils.py#L245-L248

File: pydis_site/apps/content/urls.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/urls.py]: It wasn't intentional, I'll see about removing it.

File: pydis_site/apps/content/views/tags.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/views/tags.py]: Hmm, that dash is for tags, not tag groups, so I think it's still necessary (for instance `args-kwargs`). I hadn't considered tag groups at all when writing this, but I don't think they're separated by spaces either, I think the tag code does not support that. The relevant code for parsing tag commands is:

https://github.com/python-discord/bot/blob/dc9cb7910040f32d91aa546eec23b942a5a72203/bot/exts/info/tags.py#L66-L68


which leads me to believe that tags currently have to follow the `(tag-group)? tag-name` structure. I think it's simplest to write a regex which captures everything after `!tag` and handle the logic in the replacement function.

File: pydis_site/apps/content/utils.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/utils.py:20]: What would we apply the cache to here? The code checks the TTL against items from the database.

File: pydis_site/apps/content/models/tag.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/models/tag.py]: There's also the option of just not caring, it's unusual for the commit author to not be the committer as well, so it's not very important info to display.

File: pydis_site/apps/content/models/tag.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/models/tag.py]: Yeah that's a fair point, c0384c62

File: pydis_site/apps/content/models/tag.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/models/tag.py]: c0384c62

File: pydis_site/apps/content/urls.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/urls.py]: c0384c62

File: pydis_site/apps/content/utils.py
Comment (ChrisLovering) [reply pydis_site/apps/content/utils.py:20]: Ah, this was my missunderstanding, I thought the ttl was used for an lru cache, but now I see what you mean. Ignore this.

File: pydis_site/apps/content/views/tags.py
Comment (ChrisLovering) [reply pydis_site/apps/content/views/tags.py]: Yea that seems reasonable, when i said separated by spaces I meant the group and the tag itself, I forgot tags could have hyphens in them too.

File: pydis_site/apps/content/models/tag.py
Comment (ChrisLovering) [reply pydis_site/apps/content/models/tag.py]: yea I don't think we need to go as far as another model, I was just surprised when reviewing that a field titled `author` was actually a serialized json structure _possibly_ containing multiple authors with additional info.

File: pydis_site/apps/content/views/tags.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/views/tags.py]: Actually it might be pretty difficult to accurately discern what's a tag group, and what's a tag with a normal sentence after it (for instance this tag link at the bottom of [args-kwargs](https://deploy-preview-763--pydis-static.netlify.app/pages/tags/args-kwargs/) which has a sentence after it). The only reliable way would be to either make a request to the page (reverse does not validate kwargs), or to use the get_tag function, but that can make an extra request (commit data) if the tag is not part of a group, neither of which are ideal solutions.

Comment (HassanAbouelela) [reply pydis_site/apps/content/views/tags.py]: This regex achieves the first part of that:
```regex
`*!tags? (?P<first>[\w-]+)(?P<second> [\w-]+)?`*
```

It might optionally match a second word after the first one, and if it does, that could be a tag group combination

Comment (ChrisLovering) [reply pydis_site/apps/content/views/tags.py]: Seems like it would do the job, obviously we'd need to update the tag model and the logic to get the url. So I'd be happy to merge this and have that as a separate PR if needed. Especially since we don't have any tag groups atm anyway.

**Review by ChrisLovering — APPROVED** [review]

File: pydis_site/apps/content/views/tags.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/views/tags.py]: I wrote the code for it, that was pretty simple, but I'm having unrelated trouble getting the test suite running. I'll probably have the fix out tomorrow

Comment (HassanAbouelela) [reply pydis_site/apps/content/views/tags.py]: Fixed in efe784fe

File: pydis_site/apps/content/utils.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/utils.py:20]: Fair enough, thanks for the library suggestion though, it looks cool 😀

File: pydis_site/apps/content/models/tag.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/models/tag.py]: 2343f7ab

File: pydis_site/apps/content/urls.py
Comment (jchristgit) [inline pydis_site/apps/content/urls.py:45]: A comment couldn't hurt here as to why we start with `{None}`. I can understand it, as Chris has lovingly explained it to me.

File: pydis_site/apps/content/utils.py
Comment (jchristgit) [inline pydis_site/apps/content/utils.py]: Why not use `with`?

File: pydis_site/apps/content/utils.py
Comment (jchristgit) [inline pydis_site/apps/content/utils.py]: This part does not make sense to me. If the hash of the file changes, shouldn't we update it here? We don't seem to throw it away in that case either, but just stop updating.

File: pydis_site/apps/content/utils.py
Comment (jchristgit) [inline pydis_site/apps/content/utils.py:196]: This typehint is wrong, because `return Tag.objects.all()` returns a `QuerySet`, **not** a `list[Tag]`. We could use `Iterable[Tag]` or possibly even `Container[Tag]`, but `list[Tag]` is wrong.

File: pydis_site/apps/content/utils.py
Comment (jchristgit) [inline pydis_site/apps/content/utils.py]:
```suggestion
        group, name = path
```

File: pydis_site/apps/content/utils.py
Comment (jchristgit) [inline pydis_site/apps/content/utils.py]: Use `str.casefold()` instead of `.lower()` for key functions.

```suggestion
    return {tag["title"]: tag for tag in sorted(data, key=lambda tag: tag["title"].casefold())}
```

Comment (jchristgit) [reply pydis_site/apps/content/utils.py]: This may cause one of the 200 dictator linters to throw up due to line length.

File: pydis_site/apps/content/views/tags.py
Comment (jchristgit) [inline pydis_site/apps/content/views/tags.py:19]: This is a great comment, and I love it very much.

**Review by jchristgit — CHANGES_REQUESTED** [review]
Legend man. LEGEND!

I have small comments. If the comments are introspected, I am approving.

File: pydis_site/apps/content/utils.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/utils.py]: We don't fetch the hashes here as documented in the approach methodology in the PR body since it would be too expensive in terms of ratelimits. There are ~100 tags, and it doesn't make sense to make an additional 100 requests to pre-load all the commits. When someone requests a tag, that commit data is fetched and saved, otherwise no commit data will be available until then.

**Review by jchristgit — APPROVED** [review]
Great Behaviour From Joe Banks In The Dog Park Today.

File: pydis_site/apps/content/views/tags.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/views/tags.py:19]: Thank you :D

File: pydis_site/apps/content/utils.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/utils.py]: 426fc71b

File: pydis_site/apps/content/utils.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/utils.py:196]: 153059c

File: pydis_site/apps/content/utils.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/utils.py]: e8b3745c

File: pydis_site/apps/content/urls.py
Comment (HassanAbouelela) [reply pydis_site/apps/content/urls.py:45]: 3940e34b

**Pull request merged**
